presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java

```diff
@@ -0,0 +1,513 @@
+    {
+        OGCGeometry geometry = deserialize(input);
+        Envelope envelope = getEnvelope(geometry);
+        return envelope.getYMax();
+    }
+
+    @Description("Returns the minimum X coordinate of geometry")
+    @ScalarFunction
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stMinX(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        Envelope envelope = getEnvelope(geometry);
+        return envelope.getXMin();
+    }
+
+    @Description("Returns the minimum Y coordinate of geometry")
+    @ScalarFunction
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stMinY(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        Envelope envelope = getEnvelope(geometry);
+        return envelope.getYMin();
+    }
+
+    @Description("Returns the number of interior rings in the polygon")
+    @ScalarFunction
+    @SqlType(StandardTypes.BIGINT)
+    public static long stInteriorRingNumber(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "st_interior_ring_number", POLYGON);
+        return ((OGCPolygon) geometry).numInteriorRing();
+    }
+
+    @Description("Returns the number of points in the geometry")
+    @ScalarFunction
+    @SqlType(StandardTypes.BIGINT)
+    public static long stPointNumber(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        if (geometry.geometryType().equals(POLYGON)) {
+            Polygon polygon = (Polygon) geometry.getEsriGeometry();
+            return polygon.getPointCount() + polygon.getPathCount();
+        }
+        else if (geometry.geometryType().equals(POINT)) {
+            return geometry.getEsriGeometry().isEmpty() ? 0 : 1;
+        }
+        else if (geometry.geometryType().equals(MULTIPOINT)) {
+            return ((MultiPoint) geometry.getEsriGeometry()).getPointCount();
+        }
+        return ((MultiPath) geometry.getEsriGeometry()).getPointCount();
+    }
+
+    @SqlNullable
+    @Description("Returns true if and only if the line is closed and simple")
+    @ScalarFunction
+    @SqlType(StandardTypes.BOOLEAN)
+    public static Boolean stIsRing(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "st_is_ring", LINE);
+        OGCLineString line = (OGCLineString) geometry;
+        return line.isClosed() && line.isSimple();
+    }
+
+    @Description("Returns the start point of an line")
+    @ScalarFunction
+    @SqlType(GEOMETRY)
+    public static Slice stStartPoint(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "st_start_point", LINE);
+        MultiPath lines = (MultiPath) geometry.getEsriGeometry();
+        SpatialReference reference = geometry.getEsriSpatialReference();
+        return serialize(OGCGeometry.createFromEsriGeometry(lines.getPoint(0), reference));
```

Comment:
Why do we need the SpatialReference? shouldn't this just return `lines.getPoint(0)`? (See my comment above regarding euclidean geometry vs geospatial)

Reply:
need to serialize an OGCGeometry, Point is not OGCGeometry